core/unix/memquery_test.h

```diff
@@ -0,0 +1,311 @@
+    return true;
+}
+
+bool
+memquery_iterator_next_test(memquery_iter_t *iter)
+{
+    if (!enable_memquery_unit_test) {
+        return memquery_iterator_next(iter);
+    }
+    if (next_iter_position >= cur_bounds.iters_count) {
+        return false;
+    }
+    *iter = cur_bounds.iters[next_iter_position++].iter_result;
+    return true;
+}
+
+void
+memquery_iterator_stop_test(memquery_iter_t *iter)
+{
+    if (!enable_memquery_unit_test) {
+        return memquery_iterator_stop(iter);
+    }
+}
+
+bool
+module_is_header_test(app_pc base, size_t size)
+{
+    if (!enable_memquery_unit_test) {
+        return module_is_header(base, size);
+    }
+    if (cur_bounds.iters_count == 0) {
+        print_file(STDERR, "module_is_header_test called without setting cur_bounds!");
+        *(int *)(42) = 43;
+    }
+    for (int i = 0; i < cur_bounds.iters_count; i++) {
+        fake_memquery_result *cur = &cur_bounds.iters[i];
+        memquery_iter_t iter = cur->iter_result;
+        if (iter.vm_start <= base && base < iter.vm_end) {
+            return cur->is_header;
+        }
+    }
+    print_file(STDERR, "UNKNOWN BASE PC " PFX "\n", base);
+    return false;
+}
+
+bool
+module_walk_program_headers_test(app_pc base, size_t view_size, bool at_map,
+                                 bool dyn_reloc, OUT app_pc *out_base /* relative pc */,
+                                 OUT app_pc *out_first_end /* relative pc */,
+                                 OUT app_pc *out_max_end /* relative pc */,
+                                 OUT char **out_soname, OUT os_module_data_t *out_data)
+{
+    if (!enable_memquery_unit_test) {
+        return module_walk_program_headers(
+            base, view_size, at_map, dyn_reloc, out_base /* relative pc */,
+            out_first_end /* relative pc */, out_max_end /* relative pc */, out_soname,
+            out_data);
+    }
+    if (cur_bounds.iters_count == 0) {
+        ASSERT(false && "module_is_header_test called without setting cur_bounds!");
+    }
+    if (out_first_end != NULL || out_soname != NULL || out_data != NULL ||
+        out_base == NULL || out_max_end == NULL) {
+        ASSERT(false &&
+               "out_data, out_first_end, and out_soname must be NULL, and out_base and "
+               "out_max_end must not be!");
+    }
+    for (int i = 0; i < cur_bounds.iters_count; i++) {
+        fake_memquery_result *cur = &cur_bounds.iters[i];
+        memquery_iter_t iter = cur->iter_result;
+        if (iter.vm_start <= base && base < iter.vm_end) {
+            if (cur->is_header) {
+                *out_base = cur->mod_base;
+                *out_max_end = cur->mod_end;
+                return true;
+            }
+        }
+    }
+    ASSERT(false && "UNKNOWN BASE PC");
+    return false;
+}
+
+/* record_memquery_library_bounds_by_iterator wraps
+ * memquery_library_bounds_by_iterator, printing out test case data for the
+ * memquery unit test. See memquery_test_data.h for a selection of tests.
```

Comment:
I added a note in memquery_test_data.h to point to record_memquery_library_bounds_by_iterator.

I should have included more instructions, thanks for asking: my code evolved a little between the time I wrote the record function and the unit tests, and things do integrate well together in an obvious way. I've included some notes on how to get this code to automatically output test cases but it is currently a manual process with a handful of manual edits to the code; I am sure I can figure out a way to do it based on a macro define instead, but I want to focus on making the unit test work with a real interface instead of basing it off #defines first. I'm leaving this open to revisit when that's done.

Reply:
Alright, not only did I include instructions on how to get this to work, I added support for automatically generating the data based on whether RECORD_MEMQUERY is defined. This should be much easier to use.